### PR TITLE
Add more details about the secure option

### PR DESCRIPTION
**Pull request by danmactough**
I've had to track down this information more than once, and since it cuts across both Express and Koa, I think the docs for this library are a good place to discover that there's another setting in your app that you need to configure if you want to use secure cookies behind an SSL-terminating proxy.

### CODE REVIEW

**Comment by dougwilson** [comment]
Awesome,thank you!

Many other frameworks besides Express and Koa use this module. If they are going to get a specific call out, can you add any other applicable frameworks you can find as well? Or at least explain how this module understands if the request is secure or not, so people can figure it out.

The option also seems a bit wordy now. Can you split this out into it's own section that I easier to read (you can add a link yo that section of course)?

**Comment by danmactough** [comment]
> Can you split this out into it's own section...?

You bet. 👍 

> can you add any other applicable frameworks you can find as well?

Yeah, I can see if I can find any others. Maybe in that new section ☝️ we can invite people to add links to config details for other frameworks?

**Comment by dougwilson** [comment]
> Maybe in that new section ☝️ we can invite people to add links to config details for other frameworks?

Exactly :) I think adding details of how this module determines of the "req" is secure would probably go along way for the others, since that way people will understand it at the level this module does.

**Comment by dougwilson** [comment]
FWIW, express itself does not actually use this module, incase there was confusion around that.

**Comment by danmactough** [comment]
Right. I think I've seen it transitively via https://github.com/expressjs/session (mostly been using koa and restify lately).

**Comment by dougwilson** [comment]
> I think I've seen it transitively via https://github.com/expressjs/session

That module doesn't use this module, either :)

**Comment by danmactough** [comment]
😲 that's my sign to call it a day

**Comment by dougwilson** [comment]
:O anyhow, I really look forward to the tweaks! This module's docs a not terrible but not amazing, either, so definitely can use some more expansion!

**Comment by danmactough** [comment]
Cool cool. Yeah, this is a particularly tricky thing ... as you are painfully aware from the troubleshooting and questions you've fielded on this topic. Like, so tricky that I bothered to [blog about it](http://blog.mact.me/2016/09/12/so-you-want-to-keep-your-cookies-secure). 😄

**Comment by danmactough** [comment]
@dougwilson I broke up that big chunk of a bullet point and explained how the library determines whether a request is secure.

I also looked at a couple other frameworks: geddy, hapi, restify. Hapi is kind of its own universe, and Geddy and Restify didn't seem to have any way to tell the framework how to trust the upstream proxy (although it would be trivial to write a middleware to make it work in Restify).

**Comment by dougwilson** [comment]
Awesome, much easier to read :) And thanks for taking a look into those other frameworks! I only have a couple comments:

1. We probably don't need a link to the source code, especially since that link it to the current `master` branch, which will be annoying to remember to keep up-to-date every time the lines are shuffled around.
2. The link text down to the section from the option above is kind of weird--I was confused what clicking on it would do when looking at the readme. Could probably just be as simple as "read more about this option" or something.

**Comment by danmactough** [comment]
> The link text down to the section from the option above is kind of weird--I was confused what clicking on it would do when looking at the readme. Could probably just be as simple as "read more about this option" or something.

Not sure I understood what you wanted, but I gave it a shot.

**Comment by dougwilson** [comment]
The change looks good enough to me 👍

File: README.md [hunk 1]
```diff
@@ -67,11 +67,29 @@ If the _options_ object is provided, it will be used to generate the outbound co
...
 * `signed`: a boolean indicating whether the cookie is to be signed (`false` by default). If this is true, another cookie of the same name with the `.sig` suffix appended will also be sent, with a 27-byte url-safe base64 SHA1 value representing the hash of _cookie-name_=_cookie-value_ against the first [Keygrip](https://www.npmjs.com/package/keygrip) key. This signature key is used to detect tampering the next time a cookie is received.
 * `overwrite`: a boolean indicating whether to overwrite previously set cookies of the same name (`false` by default). If this is true, all cookies set during the same request with the same name (regardless of path or domain) are filtered out of the Set-Cookie header when setting this cookie.
 
+### secure cookies
+
+To send a secure cookie, you set a cookie with the `secure: true` option.
+
+HTTPS is necessary for secure cookies. If you try to send a secure cookie over HTTP, the cookie will not be set. In fact, this library will **throw an error** -- which may crash your server -- if you try to send a secure cookie over HTTP.
```
Comment (dougwilson) [inline README.md]: > In fact, this library will **throw an error** -- which may crash your server -- if you try to send a secure cookie over HTTP. 

So reading this, this almost sounds like the documentation is making fun of the library it is documenting, or at least it's being cheeky to itself. Can you change this to something more stating the fact? Perhaps the following?

> HTTPS is necessary for secure cookies. When `cookies.set` is called with `secure: true` and a secure connection is not detected, an error will be thrown.

File: README.md [hunk 2]
```diff
@@ -67,11 +67,29 @@ If the _options_ object is provided, it will be used to generate the outbound co
...
+
+This library will test each request to see if it's secure by checking:
+
+* if the request protocol is "HTTPS" or
```
Comment (dougwilson) [inline README.md]: It is probably non-obvious how one would like that the request protocol is "HTTPS" here; probably say

> if the `protocol` property of the request is set to `'https'`

which matches up to how the next line reads.

File: README.md [hunk 3]
```diff
@@ -67,11 +67,29 @@ If the _options_ object is provided, it will be used to generate the outbound co
...
+* if the `connection.encrypted` property of the request is set to `true`.
+
+If your server is running behind a proxy and you are using `secure: true`, you need to configure your server to trust request headers added by your proxy when determining whether a request is using a secure connection. If your server is properly configured, `connection.encrypted` property of the request will be `true` for requests that have the right proxy headers.
```
Comment (dougwilson) [inline README.md]: > you need to configure your server to trust request headers added by your proxy when determining whether a request is using a secure connection

should maybe be

> you need to configure your server to read the request headers added by your proxy to determine whether the request is using a secure connection

Comment (dougwilson) [inline README.md]: >  If your server is properly configured, `connection.encrypted` property of the request will be `true` for requests that have the right proxy headers.

I've never seen this behavior before. What library is doing that?

**Comment by dougwilson** [comment]
Hi @danmactough any thoughts on my comments above?

File: README.md (hunk 1 above)
Comment (danmactough) [inline README.md]: Definitely didn't intend to be making fun. It is pretty surprising for a library to throw a runtime error like this, though, so I wanted to highlight it specifically.

I think your language is better.

File: README.md (hunk 3 above)
Comment (danmactough) [inline README.md]: Whoa. Good catch -- that is false. I'll just remove that sentence. 😬

Comment (danmactough) [inline README.md]: Actually, I will describe what express and koa do (which is to use the upstream proxy's protocol as `req.protocol`).

**Comment by danmactough** [comment]
@dougwilson Thanks for the ping. Made those changes you suggested.
